File: auth-center/auth-center-domain/src/test/java/com/jmsoftware/maf/authcenter/remote/impl/OssCenterFeignServiceImplTest.java
package com.jmsoftware.maf.authcenter.remote.impl;

import com.jmsoftware.maf.authcenter.remote.OssCenterFeignClient;
import com.jmsoftware.maf.common.bean.ResponseBodyBean;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

/**
 * <h1>OssCenterFeignServiceImplTest</h1>
 * Description: OssCenterFeignServiceImplTest, change description here.
 * <p>
 * <h2>Mockito JUnit 5 Extension</h2>
 * <p>There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.</p>
 * <p><strong>Pros:</strong></p>
 * <ul>
 * <li>No need to call <code>MockitoAnnotations.openMocks()</code></li>
 * <li>Validates framework usage and detects incorrect stubbing</li>
 * <li>Easy to create mocks</li>
 * <li>Very readable</li>
 *
 * </ul>
 * <p><strong>Cons:</strong></p>
 * <ul>
 * <li>Need an extra dependency on <code>org.mockito:mockito-junit-jupiter</code>, which has been included by Spring.
 * So we don&#39;t have to worry about this.</li>
 * </ul>
 *
 * @author Johnny Miller (鍾俊), e-mail: johnnysviva@outlook.com, date: 4/4/2022 8:40 AM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito With JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito With JUnit 5</a>
 **/
@Slf4j
@ExtendWith(MockitoExtension.class)
@Execution(ExecutionMode.CONCURRENT)
class OssCenterFeignServiceImplTest {
    @InjectMocks
    private OssCenterFeignServiceImpl ossCenterFeignService;
    @Mock
    private OssCenterFeignClient ossCenterFeignClient;

    @BeforeEach
    void setUp() {
        log.info("{} setUp", this.getClass().getSimpleName());
    }

    @AfterEach
    void tearDown() {
        log.info("{} tearDown", this.getClass().getSimpleName());
    }

    @Test
    void uploadSingleResource() {
        when(this.ossCenterFeignClient.uploadSingleResource(any())).thenReturn(ResponseBodyBean.ofSuccess());
        assertThrows(RuntimeException.class, () -> this.ossCenterFeignService.uploadSingleResource(null));
        log.info("Pass: PermissionServiceImplTest#uploadSingleResource");
    }
}

File: auth-center/auth-center-domain/src/test/kotlin/com/jmsoftware/maf/authcenter/remote/impl/OssCenterFeignServiceImplTest.kt
package com.jmsoftware.maf.authcenter.remote.impl

import com.jmsoftware.maf.authcenter.remote.OssCenterFeignClient
import com.jmsoftware.maf.common.bean.ResponseBodyBean
import com.jmsoftware.maf.common.domain.osscenter.write.ObjectResponse
import com.jmsoftware.maf.common.util.logger
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.parallel.Execution
import org.junit.jupiter.api.parallel.ExecutionMode
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.mock.web.MockMultipartFile

/**
 * # OssCenterFeignServiceImplTest
 *
 * Description: OssCenterFeignServiceImplTest, change description here.
 *
 * ## Mockito JUnit 5 Extension
 *
 * There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.
 *
 * **Pros:**
 *
 *  * No need to call `MockitoAnnotations.openMocks()`
 *  * Validates framework usage and detects incorrect stubbing
 *  * Easy to create mocks
 *  * Very readable
 *
 * **Cons:**
 *
 *  * Need an extra dependency on `org.mockito:mockito-junit-jupiter`, which has been included by Spring.
 * So we don&#39;t have to worry about this.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/10/22 12:38 PM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito with JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito with JUnit 5</a>
 */
@ExtendWith(MockitoExtension::class)
@Execution(ExecutionMode.CONCURRENT)
internal class OssCenterFeignServiceImplTest {
    companion object {
        private val log = logger()
    }

    @InjectMocks
    private lateinit var ossCenterFeignService: OssCenterFeignServiceImpl

    @Mock
    private lateinit var ossCenterFeignClient: OssCenterFeignClient

    @BeforeEach
    fun setUp() {
        log.info("${this.javaClass.simpleName} setUp")
    }

    @AfterEach
    fun tearDown() {
        log.info("${this.javaClass.simpleName} tearDown")
    }

    @Test
    fun uploadSingleResource() {
        val multipartFile = MockMultipartFile(
            "name-for-unit-test.txt",
            "original-filename.txt",
            null,
            null
        )
        `when`(ossCenterFeignClient.uploadSingleResource(multipartFile))
            .thenReturn(ResponseBodyBean.ofSuccess(ObjectResponse()))
        val objectResponse = ossCenterFeignService.uploadSingleResource(multipartFile)
        assertNotNull(objectResponse)
        log.info("Pass: PermissionServiceImplTest#uploadSingleResource. $objectResponse")
    }
}
